runner/status_rw.go

```diff
@@ -0,0 +1,116 @@
+package runner
+
+import (
+	"fmt"
+	"time"
+)
+
+// status_rw.go is how to read/write RunStatus'es
+
+// StateMask describes a set of States as a bitmask.
+type StateMask uint64
+
+// Useful StateMask constants
+const (
+	UNKNOWN_MASK    StateMask = StateMask(1 << uint(UNKNOWN))
+	PENDING_MASK              = 1 << uint(PENDING)
+	PREPARING_MASK            = 1 << uint(PREPARING)
+	RUNNING_MASK              = 1 << uint(RUNNING)
+	COMPLETE_MASK             = 1 << uint(COMPLETE)
+	FAILED_MASK               = 1 << uint(FAILED)
+	ABORTED_MASK              = 1 << uint(ABORTED)
+	TIMEDOUT_MASK             = 1 << uint(TIMEDOUT)
+	BADREQUEST_MASK           = 1 << uint(BADREQUEST)
+	DONE_MASK                 = (1<<uint(COMPLETE) |
+		1<<uint(FAILED) |
+		1<<uint(ABORTED) |
+		1<<uint(TIMEDOUT))
+	ALL_MASK = math.MaxUint64
+)
+
+// Helper Function to create StateMask that matches exactly state
+func MaskForState(state ProcessState) StateMask {
+	return 1 << uint(state)
+}
+
+// Query describes a query for RunStatuses.
+// The Runs and States are and'ed: a RunStatus matches a Query
+// if its ID is in q.Runs (or q.AllRuns) and its state is in q.States
+type Query struct {
+	Runs    []RunID   // Runs to query for
+	AllRuns bool      // Whether to match all runs
+	States  StateMask // What States to match
+}
+
+// Wait describes how to Wait.
+// It differs from StatusQuery because StatusQuery describes what RunStatus'es to match, but
+// Wait describes how to Wait
+type Wait struct {
+	// How long to wait for Statuses
+	Timeout time.Duration
+
+	// We might add whether to return as soon as one status matches, or waiting until all
+}
+
+// StatusQuerier allows reading Status by Query'ing.
+type StatusQuerier interface {
```

Comment:
Now that I think about it, I think eventually it will just be StatusReader, once we've shaken out the LegacyStatusReader usages and can collapse them...